ocaml-lsp-server/src/compl.ml

```diff
@@ -119,62 +115,124 @@ let range_prefix (lsp_position : Position.t) prefix : Range.t =
-  let completion =
-    let complete =
-      Query_protocol.Complete_prefix (prefix, position, [], false, true)
-    in
-    Query_commands.dispatch pipeline complete
+let complete_prefix doc prefix pos =
+  let open Fiber.O in
+  let position = Position.logical pos in
+  let+ (completion : Query_protocol.completions) =
+    (* TODO: can't we use [Document.dispatch_exn] instead of
+       [Document.with_pipeline_exn] for conciseness? *)
+    Document.with_pipeline_exn doc (fun pipeline ->
+        let complete =
+          Query_protocol.Complete_prefix (prefix, position, [], false, true)
+        in
+        Query_commands.dispatch pipeline complete)
   in
-  let short_range =
-    range_prefix lsp_position
+  let range =
+    range_prefix pos
       (prefix_of_position ~short_path:true (Document.source doc) position)
   in
-  let items =
-    completion.entries |> List.map ~f:(fun entry -> (short_range, entry))
-  in
-  let items =
+  let completion_entries =
     match completion.context with
-    | `Unknown -> items
+    | `Unknown -> completion.entries
     | `Application { Query_protocol.Compl.labels; argument_type = _ } ->
-      items
+      completion.entries
       @ List.map labels ~f:(fun (name, typ) ->
-            ( short_range
-            , { Query_protocol.Compl.name
-              ; kind = `Label
-              ; desc = typ
-              ; info = ""
-              ; deprecated = false (* TODO this is wrong *)
-              } ))
+            { Query_protocol.Compl.name
+            ; kind = `Label
+            ; desc = typ
+            ; info = ""
+            ; deprecated = false (* TODO this is wrong *)
+            })
   in
-  let textDocument = TextDocumentIdentifier.create ~uri:(Document.uri doc) in
-  let compl_info =
-    CompletionParams.create ~textDocument ~position:lsp_position ()
+  (* we need to json-ify completion params to put them in completion item's
+     [data] field to keep it across [textDocument/completion] and the following
+     [completionItem/resolve] requests *)
+  let compl_params =
+    let textDocument = TextDocumentIdentifier.create ~uri:(Document.uri doc) in
+    CompletionParams.create ~textDocument ~position:pos ()
     |> CompletionParams.yojson_of_t
   in
-  let items = List.mapi ~f:(item ~compl_info) items in
-  `CompletionList { CompletionList.isIncomplete = false; items }
+  List.mapi completion_entries
+    ~f:(completionItem_of_completion_entry ~range ~compl_params)
+
+let construct doc position =
+  let command = Query_protocol.Construct (position, None, None) in
+  let open Fiber.O in
+  let+ r = Document.dispatch doc command in
+  match r with
+  | Error { exn = Merlin_analysis.Construct.Not_a_hole; _ } -> Error `Not_a_hole
+  | Error e -> Error (`Exn e)
+  | Ok (loc, exprs) ->
+    let range = Range.of_loc loc in
+    List.mapi exprs ~f:(fun ix expr ->
+        let textEdit = `TextEdit { TextEdit.range; newText = expr } in
+        let preselect =
+          if ix = 0 then
```

Comment:
Should just be `let preselect = ix = 0 in`.

Better yet is to handle this tweak after generating the list:

```ocaml
let items = List.mapi exprs ~f:(...) in
match items with
| x :: xs -> { x with preselect = true } :: xs
| [] -> []
```

This is not for optimization, I just find adding the a check that will only fire on the first element to be distasteful.